src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs

```diff
@@ -7,187 +7,214 @@
 using System.Collections.Immutable;
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
-using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Remote;
 using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.SourceGeneration;
 using Microsoft.VisualStudio.Language.Intellisense;
 using Microsoft.VisualStudio.Shell;
 using Roslyn.Utilities;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer;
+
+internal abstract partial class BaseDiagnosticAndGeneratorItemSource : IAttachedCollectionSource
 {
-    internal abstract partial class BaseDiagnosticAndGeneratorItemSource : IAttachedCollectionSource
+    private static readonly DiagnosticDescriptorComparer s_comparer = new();
+
+    private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+    private readonly IAsynchronousOperationListener _listener;
+    private readonly BulkObservableCollection<BaseItem> _items = new();
+
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+    private readonly AsyncBatchingWorkQueue _workQueue;
+    private readonly IThreadingContext _threadingContext;
+
+    protected Workspace Workspace { get; }
+    protected ProjectId ProjectId { get; }
+    protected IAnalyzersCommandHandler CommandHandler { get; }
+
+    /// <summary>
+    /// The analyzer reference that has been found. Once it's been assigned a non-null value, it'll never be assigned
+    /// <see langword="null"/> again.
+    /// </summary>
+    private AnalyzerReference? _analyzerReference_DoNotAccessDirectly;
+
+    public BaseDiagnosticAndGeneratorItemSource(
+        IThreadingContext threadingContext,
+        Workspace workspace,
+        ProjectId projectId,
+        IAnalyzersCommandHandler commandHandler,
+        IDiagnosticAnalyzerService diagnosticAnalyzerService,
+        IAsynchronousOperationListenerProvider listenerProvider)
     {
-        private static readonly DiagnosticDescriptorComparer s_comparer = new DiagnosticDescriptorComparer();
+        _threadingContext = threadingContext;
+        Workspace = workspace;
+        ProjectId = projectId;
+        CommandHandler = commandHandler;
+        _diagnosticAnalyzerService = diagnosticAnalyzerService;
+
+        _listener = listenerProvider.GetListener(FeatureAttribute.SourceGenerators);
```

Comment:
>_listener [](http://example.com/codeflow?start=8&length=9)

nit: doesn't look like it needs to be a member

Reply:
probably not.  if so, can remove in followup if there are no other changes i need to make.